services/src/main/java/org/keycloak/protocol/oidc/par/endpoints/ParEndpoint.java

```diff
@@ -0,0 +1,201 @@
+import org.keycloak.protocol.oidc.utils.AuthorizeClientUtil;
+import org.keycloak.protocol.oidc.par.ParResponse;
+import org.keycloak.representations.idm.OAuth2ErrorRepresentation;
+import org.keycloak.services.CorsErrorResponseException;
+import org.keycloak.services.ErrorResponseException;
+import org.keycloak.services.resource.RealmResourceProvider;
+import org.keycloak.services.resources.Cors;
+import org.keycloak.utils.ProfileHelper;
+
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.HashMap;
+import java.util.Map;
+
+
+/**
+ * Pushed Authorization Request endpoint
+ */
+public class ParEndpoint implements RealmResourceProvider {
+
+    private static final Logger LOG = Logger.getLogger(ParEndpoint.class);
+    private static final String REQUEST_URI_TEMPLATE = "urn:ietf:params:oauth:request_uri:%s";
+
+    private KeycloakSession session;
+
+    @Context
+    private HttpRequest request;
+
+    @Context
+    private HttpResponse httpResponse;
+
+    @Context
+    private HttpHeaders headers;
+
+    @Context
+    private ClientConnection clientConnection;
+
+    private EventBuilder event;
+    private Cors cors;
+    private RealmModel realm;
+
+    public ParEndpoint(KeycloakSession session) {
+        this.session = session;
+    }
+
+    @POST
+    @Consumes({MediaType.APPLICATION_FORM_URLENCODED})
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response handlePar() {
+
+        ProfileHelper.requireFeature(Profile.Feature.PAR);
+
+        LOG.debug("Received PAR object");
+
+        KeycloakContext context = session.getContext();
+        realm = context.getRealm();
+        event = new EventBuilder(realm, session, context.getConnection()).event(EventType.PUSHED_AUTHORIZATION_REQUEST);
+
+        checkSsl();
+        checkRealm();
+        authorizeClient();
+
+        ClientModel clientModel = session.getContext().getClient();
+
+        ParValidationService parValidationService = new ParValidationService(session, event);
+
+        Response validationResponse = parValidationService.validateParRequest(request, clientModel);
+
+        if (validationResponse != null) {
+            return validationResponse;
+        }
+
+        Map<String, String> params = new HashMap<>();
+
+        byte[] clientHashAndSecret = Bytes.concat(getHash(clientModel.getClientId()), KeycloakModelUtils.generateSecret());
+        String requestUri = String.format(REQUEST_URI_TEMPLATE, Base64Url.encode(clientHashAndSecret));
+
+        int expiresIn = realm.getAttribute("requestUriLifespan", 60);
+
+        request.getFormParameters().forEach((k, v) -> params.put(k, String.valueOf(v)));
```

Comment:
`request.getFormParameters()` returns percent encoded redirect_uri.
`request.getDecodedFormParameters()` may resolve the redirect_uri related problem you've encountered.

Reply:
@tnorimat this is not the issue, let me please describe it more precisely. First, I have provided redirect URI in my client configuration - see the picture 1:

<img width="569" alt="Screen Shot 2021-06-02 at 3 28 33 PM" src="https://user-images.githubusercontent.com/64079106/120480042-4d679e80-c3b7-11eb-94f8-f71df039c96a.png">

After launching the whole flow and providing the user's credentials, I can see the following in the browser console (picture 2):

![wrong1](https://user-images.githubusercontent.com/64079106/120480202-75ef9880-c3b7-11eb-855a-778aa9b1e913.png)

That is why this request fails with 404 and the error “We are sorry…“ is displayed. Please note, that the link is built in a **proper** way (without the "http://localhost:8081/auth/" prefix) in case of providing the redirect URI parameter "localhost:8088", even if there is no application on my machine at the 8088 port. Is there a way to switch off the "http://localhost:8081/auth/" prefix during the link building? Thanks, have a nice day.